### ISSUE TITLE
unresponsive extension host issue with `vscode.markdown-language-features`

### ISSUE THREAD

**Issue by hikchoi** [microsoft/vscode#152494]
Issue Type: <b>Performance Issue</b>

I am getting the following log (in `Log (Window)` output) after a few seconds of lag:

```
[2022-06-18 00:31:16.247] [renderer7] [info] UNRESPONSIVE extension host: starting to profile NOW
[2022-06-18 00:31:21.306] [renderer7] [warning] UNRESPONSIVE extension host: 'vscode.markdown-language-features' took 100% of 5461.375ms, saved PROFILE here: 'file:///var/folders/hw/fw5nd0c53cq3q9g8w2m4stk40000gn/T/exthost-296d96.cpuprofile' [{"id":"vscode.markdown-language-features","total":5461167,"percentage":100}]
```

Here is the screenshot of the cpu profile:
![image](https://user-images.githubusercontent.com/1219789/174339699-854e14bf-eac2-42e4-83c8-dfefb9c665de.png)

## More context

This happens intermittently when using [Dendron](https://github.com/dendronhq/dendron) to navigate between markdown notes through wikilinks (Dendron's extended markdown syntax for linking different files, using `[[double brackets]]`).

Once the lag is resolved, it doesn't happen anymore until I close VSCode and open a new session. This doesn't happen in a deterministic manner (doesn't always happen in the same markdown file), but it happens frequent enough for it to be noticeable and affect regular use.

I am unable to test this with all extensions disabled, because disabling all extensions  would remove the ability for me to reproduce this.

Since our extension is also dealing with Markdown, I am suspecting something is conflicting with `vscode.markdown-language-features` that we are not noticing. Would love some pointers to troubleshoot this so that it could be fixed on our end if needed.

[this line](https://github.com/microsoft/vscode/blob/623f55f437ab31fa56e13acafa9b9e7d4b50c4bf/extensions/markdown-language-features/src/languageFeatures/documentLinkProvider.ts#L275) seems to be where the extension host is hanging, but the markdown file that this happens does not contain any _vanilla_ markdown links, so I am scratching my head on what is going on.

---

VS Code version: Code - Insiders 1.69.0-insider (53130238b422606ffcd616126fb8c6c81c6e4436, 2022-06-13T11:22:09.918Z)
OS version: Darwin arm64 21.3.0
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M1 Pro (8 x 24)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|8, 8, 8|
|Memory (System)|16.00GB (0.07GB free)|
|Process Argv|--crash-reporter-id d168eb34-96e1-4a4a-9d29-5c5489c6aba6|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>Process Info</summary>

```
CPU %	Mem MB	   PID	Process
   10	   131	 62990	code-insiders main
    5	    66	 62993	   gpu-process
    0	    33	 62995	   utility-network-service
   37	   131	 63340	   shared-process
    0	    16	 63343	     ptyHost
    0	    33	 91168	     fileWatcher
    0	     0	 92628	     /bin/ps -ax -o pid=,ppid=,pcpu=,pmem=,command=
    0	   115	 91167	   extensionHost
    0	    33	 91182	     /Applications/Visual Studio Code - Insiders.app/Contents/Frameworks/Code - Insiders Helper.app/Contents/MacOS/Code - Insiders Helper --ms-enable-electron-run-as-node /Users/hikchoi/.vscode-insiders/extensions/dendron.nightly-0.99.3/dist/server.js
    0	    33	 91286	     /Applications/Visual Studio Code - Insiders.app/Contents/Frameworks/Code - Insiders Helper.app/Contents/MacOS/Code - Insiders Helper --ms-enable-electron-run-as-node /Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/extensions/json-language-features/server/dist/node/jsonServerMain --node-ipc --clientProcessId=91167
    0	    33	 91736	     /Applications/Visual Studio Code - Insiders.app/Contents/Frameworks/Code - Insiders Helper.app/Contents/MacOS/Code - Insiders Helper --ms-enable-electron-run-as-node --max-old-space-size=3072 /Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js --serverMode partialSemantic --useInferredProjectPerProjectRoot --disableAutomaticTypingAcquisition --cancellationPipeName /var/folders/hw/fw5nd0c53cq3q9g8w2m4stk40000gn/T/vscode-typescript501/ac41d51e03fd359b1334/tscancellation-0b28baaed4662a5db171.tmp* --locale en --noGetErrOnBackgroundUpdate --validateDefaultNpmLocation --useNodeIpc
    0	    49	 91737	     /Applications/Visual Studio Code - Insiders.app/Contents/Frameworks/Code - Insiders Helper.app/Contents/MacOS/Code - Insiders Helper --ms-enable-electron-run-as-node --max-old-space-size=3072 /Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js --useInferredProjectPerProjectRoot --enableTelemetry --cancellationPipeName /var/folders/hw/fw5nd0c53cq3q9g8w2m4stk40000gn/T/vscode-typescript501/ac41d51e03fd359b1334/tscancellation-26e323920bf529c31251.tmp* --locale en --noGetErrOnBackgroundUpdate --validateDefaultNpmLocation --useNodeIpc
    0	    33	 91738	       /Applications/Visual Studio Code - Insiders.app/Contents/Frameworks/Code - Insiders Helper.app/Contents/MacOS/Code - Insiders Helper --ms-enable-electron-run-as-node /Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/extensions/node_modules/typescript/lib/typingsInstaller.js --globalTypingsCacheLocation /Users/hikchoi/Library/Caches/typescript/4.7 --enableTelemetry --typesMapLocation /Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/extensions/node_modules/typescript/lib/typesMap.json --validateDefaultNpmLocation
    2	   115	 91558	   window (undefined)
    8	    82	 92626	   issue-reporter
   16	   328	 97789	   window (extension.js — dendron (Workspace))
```

</details>
<details>
<summary>Workspace Info</summary>

```
|  Window (extension.js — dendron (Workspace))
|    Folder (notes): 317 files
|      File types: md(305) json(2) yml(2) gitignore(1) code-snippets(1) png(1)
|                  log(1)
|      Conf files: package.json(1)
|    Folder (notes): 606 files
|      File types: md(591) yml(7) json(1) gitignore(1) code-snippets(1) png(1)
|                  gif(1)
|      Conf files:
|    Folder (notes): 3976 files
|      File types: md(3960) yml(6) png(5) json(1) code-snippets(1)
|      Conf files:
|    Folder (notes): 30 files
|      File types: md(25) yml(2) json(1) code-snippets(1)
|      Conf files:
|    Folder (vault): 922 files
|      File types: md(884) png(24) yml(6) code-search(2) json(1)
|                  code-snippets(1) pdf(1)
|      Conf files:
|    Folder (notes): 237 files
|      File types: md(228) png(2) yml(2) json(1) code-snippets(1)
|      Conf files:
|    Folder (notes): 12579 files
|      File types: md(12495) png(51) yml(15) json(5) txt(3) gitignore(1)
|                  code-snippets(1) log(1) xlsx(1) 10(1)
|      Conf files: package.json(1)
|    Folder (notes): 322 files
|      File types: md(315) json(2) gitignore(1) code-snippets(1) yml(1)
|      Conf files: package.json(1)
|    Folder (notes): 78 files
|      File types: md(73) json(1) gitignore(1) code-snippets(1) yml(1)
|      Conf files:
|    Folder (local): 1 files
|      File types: md(1)
|      Conf files:;
```

</details>
<details><summary>Extensions (20)</summary>

Extension|Author (truncated)|Version
---|---|---
copy-relative-path|ale|0.0.2
emojisense|bie|0.9.1
dendron-markdown-shortcuts|den|0.12.1
dendron-paste-image|den|1.1.0
dendron-snippet-maker|den|0.1.6
nightly|den|0.99.3
gitlens-insiders|eam|2022.6.1705
markdown-table-formatter|fcr|2.2.4
vscode-pull-request-github|Git|0.45.2022061609
gc-excelviewer|Gra|4.2.55
vscode-peacock|joh|4.0.1
change-string-case|max|1.1.2
remote-ssh|ms-|0.83.2022061415
remote-ssh-edit|ms-|0.80.0
vscode-js-profile-flame|ms-|1.0.2
vsliveshare|ms-|1.0.5625
gremlins|nho|0.26.0
vscode-yaml|red|1.8.0
vscode-markdown-table-format|Tom|1.0.1
vim|vsc|1.22.2

(2 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv695:30137379
vsins829:30139715
vsliv368:30146709
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
pythonvspyl392:30422396
pythontb:30258533
pythonvspyt551:30291412
pythonptprofiler:30281269
vshan820:30294714
pythondataviewer:30285072
vscod805cf:30301675
bridge0708:30335490
bridge0723:30353136
vsaa593:30376534
pythonvs932:30404738
wslgetstarted:30449409
vscscmwlcmt:30465136
cppdebug:30492333
vscaat:30438846
pylanb8912:30496796
vsclangdf:30492506

```

</details>

**Comment by mjbvz** [comment]
Can you please test the next insiders build. These changes should all have made this faster:

- #152674
- #152661
- #152563
- #152533

However I don't test on Dendron workspaces so can't sure for sure

**Comment by hikchoi** [comment]
Thank you for the quick response @mjbvz 

We'll test it on our end and report back.

**Comment by hikchoi** [comment]
@mjbvz 

Just tested out again, I still see the same slowdown.

```
Version: 1.69.0-insider
Commit: 668c538b8d2fec1258ed786ae212124d6d68023c
Date: 2022-06-23T05:16:19.393Z (9 hrs ago)
Electron: 18.3.4
Chromium: 100.0.4896.160
Node.js: 16.13.2
V8: 10.0.139.17-electron.0
OS: Darwin arm64 21.3.0
```

output view: 
```
[2022-06-23 23:15:46.177] [renderer1] [info] UNRESPONSIVE extension host: starting to profile NOW
[2022-06-23 23:15:46.379] [renderer1] [info] UNRESPONSIVE extension host: received responsive event and cancelling profiling session
[2022-06-23 23:15:46.438] [renderer1] [warning] UNRESPONSIVE extension host: 'vscode.markdown-language-features' took 94% of 1060.375ms, saved PROFILE here: 'file:///var/folders/hw/fw5nd0c53cq3q9g8w2m4stk40000gn/T/exthost-bea8ba.cpuprofile' [{"id":"dendron.nightly","total":3836,"percentage":0},{"id":"eamodio.gitlens-insiders","total":333,"percentage":0},{"id":"gc","total":36958,"percentage":3},{"id":"GitHub.vscode-pull-request-github","total":501,"percentage":0},{"id":"program","total":250,"percentage":0},{"id":"self","total":15209,"percentage":1},{"id":"vscode.markdown-language-features","total":999915,"percentage":94},{"id":"vscodevim.vim","total":3207,"percentage":0}]
```

The output view says `'vscode.markdown-language-features' took 94% of 1060.375ms`, but before the output view was able to start profiling, there was an extra 2-3000ms hang.

**Comment by mjbvz** [comment]
Can you share a workspace that causes this issue?

**Comment by hikchoi** [comment]
@mjbvz 
The workspace I am seeing this issue is an internal one that I can't quite share. It is quite a large workspace with over 20000~ markdown files.

I'll try to create a test workspace of a similar size.

In the meantime, I managed to profile the exact moment when this happens in said workspace. I think it paints a better picture than the one that was automatically profiled by vscode:

Seems like github doesn't allow uploading `.cpuprofile` files. so here's a txt of it.
[perf.cpuprofile.txt](https://github.com/microsoft/vscode/files/8977530/perf.cpuprofile.txt)

**Comment by mjbvz** [comment]
Thanks yes that's a lot of MD files :)

#153395 Should improve diagnostic performance by only recomputing them for the opened files (instead of the entire workspace). However if you trigger `find all references` in some way, we will still need to scan every file in the workspace to find links to the current file. I don't think we can avoid this

Eventually moving this computation to a proper language server should make this warning go away however

**Comment by hikchoi** [comment]
I wasn't able to create a similarly sized workspace that I can reliably reproduce this issue, but looks like the perf results helped pinpointing the issue :) Thanks again for the quick response.

> https://github.com/microsoft/vscode/pull/153395 Should improve diagnostic performance by only recomputing them for the opened files (instead of the entire workspace).

I'll test it out again once the commit goes out today.

> However if you trigger find all references in some way, we will still need to scan every file in the workspace to find links to the current file. I don't think we can avoid this

I wasn't triggering any reference providing capabilities (both VSCode nor the one our extension contributes) when I encountered this slowdown, so I think we are safe :)
In case this info helps in the future, I was triggering `go to definition` on a definition for something that is not part of the vanilla markdown syntax.

> Eventually moving this computation to a proper language server should make this warning go away however

That would be awesome. Is this something planned for the future?

**Comment by hikchoi** [comment]
Just tested out with the new insiders build:
```
Version: 1.69.0-insider
Commit: e7751367f79f327b8103bee33a07c16817a2c7cd
```

The slowdown still exists unfortunately.

Is there a way to see the log outputs of `vscode.markdown-language-features`? 
I would like to take a deeper look at what is happening and find any possible interaction points with what our extension is doing.

In any case, this issue seems to only happen in a very large workspace, and only once per session (seems like there's some caching?).
For instance, in our workspace that has 20000+ markdown files the slowdown is at multiple thousands of milliseconds, but for a reasonable size (1000~2000 markdown files), that number goes down to a couple hundred milliseconds.

**Comment by mjbvz** [comment]
Markdown diagnostics are now computed in a language server so we should not block the extension host anymore

Give it a try with the latest insiders. There may still be perf issues when working with so many markdown files in a workspace but at least now it shouldn't effect other extensions

**Comment by ksdavidc** [comment]
Was this supposed to be fixd in latest install? I am still getting it.

Version: 1.69.2
Commit: 3b889b090b5ad5793f524b5d1d39fda662b96a2a
Date: 2022-07-18T16:12:57.074Z (1 wk ago)
Electron: 18.3.5
Chromium: 100.0.4896.160
Node.js: 16.13.2
V8: 10.0.139.17-electron.0
OS: Darwin x64 17.7.0Version: 1.69.2
Commit: 3b889b090b5ad5793f524b5d1d39fda662b96a2a
Date: 2022-07-18T16:12:57.074Z (1 wk ago)
Electron: 18.3.5
Chromium: 100.0.4896.160
Node.js: 16.13.2
V8: 10.0.139.17-electron.0
OS: Darwin x64 17.7.0Version: 1.69.2
Commit: 3b889b090b5ad5793f524b5d1d39fda662b96a2a
Date: 2022-07-18T16:12:57.074Z (1 wk ago)
Electron: 18.3.5
Chromium: 100.0.4896.160
Node.js: 16.13.2
V8: 10.0.139.17-electron.0
OS: Darwin x64 17.7.0

'vscode.markdown-language-features' took 98% of 581.608ms, saved PROFILE here: 'file:///var/folders/jb/10wzcs9d2d14wzw2h_26vkhw0000gn/T/exthost-9b76bd.cpuprofile' [{"id":"Durzn.brackethighlighter","total":291,"percentage":0},{"id":"ecmel.vscode-html-css","total":841,"percentage":0},{"id":"gc","total":3003,"percentage":1},{"id":"imagio.vscode-dimmer-block","total":222,"percentage":0},{"id":"naumovs.color-highlight","total":2036,"percentage":0},{"id":"oderwat.indent-rainbow","total":368,"percentage":0},{"id":"program","total":140,"percentage":0},{"id":"self","total":1580,"percentage":0},{"id":"vscode.markdown-language-features","total":572708,"percentage":98}]

**Comment by ksdavidc** [comment]
> if you trigger `find all references` in some way

And what might be doing that? This has actually been crippling me for some time, and I really want to avoid it if possible, but I need a clue. I too have a large number of md files, extensions, so hard to diagnose.

**Comment by ksdavidc** [comment]
What am I doing when the error happens? typing. just typing.

It is usually preceded by a delay of varying length, as reported above.

[exthost-9495a6.cpuprofile.txt](https://github.com/microsoft/vscode/files/9194524/exthost-9495a6.cpuprofile.txt)

**Comment by mjbvz** [comment]
@ksdavidc You're on 1.69. This should only be fixed in insiders

**Comment by mjbvz** [comment]
@hikchoi Can you please test on insiders and let me know if you're still seeing any issues

**Comment by hikchoi** [comment]
@mjbvz Unfortunately still an issue.
This does go away once the initial slowdown happens.
This only happens to very large workspaces, so it's not on the highest priority list on our end.

> Markdown diagnostics are now computed in a language server so we should not block the extension host anymore

I have a feeling that this isn't related to anything diagnostics, but I'm not super familiar with the VSCode internals so I can't say for sure. I can see a lot of tokenize calls coming from `vscode.markdown-language-features`, but not sure why. I don't think we are triggering that from our extension.

```
Version: 1.70.0-insider
Commit: b4d5b4ed69af953901051b1f049851b9dca3563e
Date: 2022-07-28T05:16:00.371Z (1 hr ago)
Electron: 18.3.5
Chromium: 100.0.4896.160
Node.js: 16.13.2
V8: 10.0.139.17-electron.0
OS: Darwin arm64 21.3.0
```

[perf-insiders.cpuprofile.txt](https://github.com/microsoft/vscode/files/9206137/perf-insiders.cpuprofile.txt)

**Comment by mjbvz** [comment]
@hikchoi Thanks. I can't repo anymore on my end though, even on larger repos. Could your extension be triggering `find all references` (possibly indirectly though `provide definition`)?

<img width="425" alt="Screen Shot 2022-07-28 at 8 39 39 AM" src="https://user-images.githubusercontent.com/12821956/181579882-21cd49fc-8fef-4b1c-9cf4-fa5c2b895b2b.png">

Right now `go to definition` is implemented using`find all references` which traverses the entire workspace (although it probably shouldn't)

**Comment by hikchoi** [comment]
@mjbvz interesting. is this from the perf I posted? I must have missed that. I'll investigate where we are triggering `provideDefinition` and get back to you. Thanks for taking the time to look into this.

**Comment by mjbvz** [comment]
Should be fixed by https://github.com/microsoft/vscode-markdown-languageservice/pull/12

Note that `find all references` still needs to pull in lots of files so that case may still be slow

**Comment by TylerLeonhardt** [comment]
@hikchoi have you seen improvement here in the insiders build?

**Comment by hikchoi** [comment]
Hey @TylerLeonhardt The issue seems to be completely resolved on my end.

Thank you for the fix! @mjbvz
